### DISCUSSION TITLE
Supported Python versions

### DISCUSSION THREAD

**Discussion by janosg** [aai-institute/pyDVL#568] [Main]
## Intro

I saw that pyDVL keeps supporting even very old Python versions. I don't have strong opinions on the topic but would like to start a discussion to hear other opinions. 

## Current status

pyDVL currently officially supports python versions 3.8 to 3.11. The policy is to keep supporting versions until they reach their end of life. I don't know if there is a specific policy for adding new versions. It seems like 3.12 would be overdue. 

## Guidelines by [Scientific Python](https://scientific-python.org)

The Scientific Python project is an initiative with the goal to better coordinate the ecosystem and support the community of contributors and maintainers. They publish guidelines for several topics. The purpose of the guidelines is to ensure quality of packages in the scientific python ecosystem but also to reduce the burden of maintainers. 

[SPEC0](https://scientific-python.org/specs/spec-0000/) details guidelines for supported Python and Dependency versions. The policy can roughly be summarized as "support the three latest Python versions". 

## How other packages handle it

- Pandas: 3.9 - 3.12
- Numpy: 3.9 - 3.12
- JAX: 3.10 - 3.12 (explicitly following SPEC0)

## Main advantages of dropping older versions

- Typically, the oldest supported version is used in the development environment. Dropping old versions means we can enjoy better error messages and speed during development
- Typing features are continually improved. 3.10 adds `|` for union types. 3.11 ads the `Self` type. 3.12 improves support for generic classes and functions.

**Comment by mdbenito** [comment]
I am not a big fan of upgrading for the sake of upgrading tbh. The new features are nice, and I sometimes would very much like to have them, but I haven't hit a situation where I severely missed one. A good argument to upgrade could be main dependencies with support for the oldest python versions reaching end of life. Numpy 1.24 (the last one to support 3.8) for instance [goes EOL on December 19th, 2024](https://endoflife.date/numpy). That would be the last good moment to drop 3.8 and require numpy 1.25. Pandas. But that's beyond EOL for python 3.8, so not much is gained...

In the end what matters is: cost / benefit of supporting older versions. Does anyone have anything to say about the cost of supporting 3.8 right now vs the benefit of not doing so? [Changes in 3.9](https://docs.python.org/3/whatsnew/3.9.html)

**Reply by mdbenito** [reply]
See #541

**Comment by mdbenito** [comment]
As of yesterday, we also support 3.9 - 3.12
